### ISSUE TITLE
 Fastify should not suppress 405 HTTP status code

### ISSUE THREAD

**Issue by addityasingh** [fastify/fastify#917]
#### You have already researched for similiar issues?
Yes

#### Are you sure this is an issue with the fastify or are you just looking for some help?

Yes. Fastify suppresses `405 Method Not Allowed` status codes and instead returns 404 

#### Is this a security related issue?
No

#### What are you trying to achieve or the steps to reproduce?

Create a server with a route handler for only `GET, POST` and write a test to validate that the server returns `405` for `HEAD` request

```js
fastify.route({
    method: ['GET', 'POST'],
    url: options.url || '/graphql',
    handler: async (req, reply) => {
      reply
        .code(200)
        .send({'Hello': 'World!'});
    },
});
```

#### What was the result you received?
404 HttpStatusCode

#### What did you expect?
405 HttpStatusCode

#### Context

* *node version*: 7,8,9
* *fastify version*: 1.2.0
* *os*: Mac
* Relevant links for reference: 
    - https://github.com/fastify/fastify/pull/862#issuecomment-385523625
    - https://github.com/addityasingh/apollo-server/blob/46b7951961f6153f84176d617d985858f3c55191/packages/apollo-server-fastify/src/fastifyApollo.ts#L72-L88

**Comment by delvedor** [comment]
This is the correct behavior, our router supports all HTTP methods, so if an handler has not been registered for a given method the answer will always be a 404 and not 405, because all HTTP methods are allowed.

If you need a workaround for this you can add an `onRequest` hook and close every incoming request with the methods you don't want to support.

```js
fastify.addHook('onRequest', (req, res, next) => {
  if (req.method === 'GET') return next()
  if (req.method === 'POST') return next()

  res.statusCode = 405
  res.setHeader('allow', 'GET, POST')
  res.end('Method Not Allowed')
})
```

**Comment by addityasingh** [comment]
@delvedor I agree with the workaround using `preRequest`. I also used the below as a workaround for now:

```javascript
fastify.route({
    method: ['HEAD', 'PUT', 'DELETE', 'OPTIONS', 'PATCH'],
    url: options.url || '/graphql',
    handler: async (req, reply) => {
      reply
        .code(405)
        .header('allow', 'GET, POST')
        .send();
    },
});
```

But based on the definition of **405** below, if I have a route definition for a resource with subset of methods and not all the methods, then I should get 405.

[405](https://httpstatuses.com/405)
> The method received in the request-line is known by the origin server but not supported by the target resource.

E.g. If I defined a route `/graphql` with allowed methods as `[GET, POST]` only, any request with methods apart from these (`PUT, HEAD, PATCH, OPTIONS` etc.) should throw a **405** because the resource (`/graphql`) is present but the method is not

```javascript
const app = require('fastify')();
  app.route({
    method: ['GET', 'POST'],
    url: options.url || '/graphql',
    handler,
  });
```

#### 2xx, 5xx or 404
```javascript
const request = require('supertest');
request(app).post('/graphql'); // This should be either 2xx or 5xx but not 4xx
request(app).get('/graphql'); // This should be either 2xx or 5xx but not 4xx
request(app).get('/graph'); // This should be a 404 since no such resource is defined
```

#### 405
```javascript
const request = require('supertest');

// These should be a 405 since resource is defined but METHOD_NOT_SUPPORTED
request(app).head('/graphql');
request(app).option('/graphql');

request(app).option('/graph'); // This again should be 404
```

Thoughts?

**Comment by nwoltman** [comment]
I tend to agree with @addityasingh. That's why I implemented automatic `405` responses in [`medley`](https://github.com/medleyjs/medley). It's not easy to implement though because this functionality is not directly supported by [`find-my-way`](https://github.com/delvedor/find-my-way).

**Comment by mcollina** [comment]
@jsumners @StarpTech @allevo what do you think?

**Comment by allevo** [comment]
I'm 👍 for responding 405 instead 404: it is more appropriate.

I have investigated but it is not simple to do.

**Comment by StarpTech** [comment]
I'm 👍 for 405 because the router capabilities don't express the user implementation. When the client request for `DELETE` but it's not available I expect a `405`.

>  The 405 (Method Not Allowed) status code indicates that the method
   received in the request-line **is known by the origin server but not
   supported by the target resource**. The origin server MUST generate an
   Allow header field in a 405 response containing a list of the target
   resource's currently supported methods.

### Some other todos:

* The origin server MUST generate an
   Allow header field in a 405 response containing a list of the target
   resource's currently supported methods.

**Comment by jsumners** [comment]
Returning a 405 would be more complete. I don't think it is necessarily *wrong* to return a 404, but the 405 definitely gives the client more information to work with. I am neutral on the change.

**Comment by mcollina** [comment]
I'm going to propose that we close this issue. I've looked into this and it is going to be very costly adding . API-wise where a 405 request will be routed? To the generic not found handle, but that's just for 404s.
Moreover this information is not readily available in the router, and it will require a major refactor of find-my-way-

I think the proposed fix:

```js
app.route({
    method: ['HEAD', 'PUT', 'DELETE', 'OPTIONS', 'PATCH'],
    url: options.url || '/graphql',
    handler: async (req, reply) => {
      reply
        .code(405)
        .header('allow', 'GET, POST')
        .send();
    },
});
```

I think we should just document this workaround.

---

This approach has been taken also by Hapi (https://github.com/hapijs/hapi/issues/3238#issuecomment-233853825):

> RFC compliance (it's a SHOULD so more like accommodation than compliance) in this case is not worth the cost. Checking for a matching route across all existing methods, just to return a slightly more accurate but really utterly useless information is not reasonable. If you care so much about this, you can manually add every possible verb to every path you configure with a 405 or 501 response. You can't do a catch all because that should remain a 404.

**Comment by mcollina** [comment]
So, we should just document this.

**Comment by addityasingh** [comment]
@mcollina I am using the workaround currently inside the `apollo-fastify-plugin` change. If we all agree that due to performance degradation that the method doesn't need to be supported, I can create a PR to add this workaround to the docs as a *Note*

**Comment by mcollina** [comment]
@addityasingh got for it, there seem to be generic agreement that supporting 405 would be too costly.

**Comment by Eomm** [comment]
@addityasingh I build a [plugin](https://github.com/Eomm/fastify-405) for this scope. I hope it could be helpful 👍

**Comment by delvedor** [comment]
Closed in #1501 🎉
